### PR TITLE
Add implementation of `OverridableBindable`

### PR DESCRIPTION
**Pull request by frenzibyte**
As in the base commit's description:
> Previously, handling of the custom difficulty values in the Difficulty Adjust mod was quite awkward, as the mod attempted to track both the base difficulty and user adjustments in a single bindable property, which could expectedly fail in some scenarios.
>
> To alleviate that, add an `OverridableBindable` more geared to this use case, that allows switching a bindable between a "base", immutable mode, and a "custom", mutable mode. The former will be used when there are no user difficulty adjustments applied, while the latter shall be used when the user enables adjusting a particular difficulty aspect.

- [ ] Depends on https://github.com/ppy/osu-framework/pull/4118 (`BindableNumberWithCurrent<T>` part)
- [ ] Depends on framework bump.

Definitely unsure of the name used here (`OverridableBindable`), that's the best I could think of.

### CODE REVIEW

File: osu.Game.Tests/NonVisual/OverridableBindableTest.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            Assert.That(setting.FinalValue.Value, Is.EqualTo(2));
+
+            parseAndEnsureExpected(7f, 7f);
+            parseAndEnsureExpected(new Bindable<float>(6f), 6f);
+            parseAndEnsureExpected(new OverridableBindable<float>(5f) { HasCustomValue = { Value = true } }, 5f);
+
+            parseAndEnsureExpected(14f, 8f);
+            parseAndEnsureExpected(-5f, 2f);
+
+            void parseAndEnsureExpected(object input, object expected)
```
Comment (bdach) [inline osu.Game.Tests/NonVisual/OverridableBindableTest.cs:96]:
```suggestion
            void parseAndEnsureExpected(object input, float expected)
```

use `object` as sparsely as humanly possible

File: osu.Game.Tests/NonVisual/OverridableBindableTest.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+            Assert.That(setting.Serialize().Deserialize<float>(), Is.EqualTo(5f));
+
+            var deserializedSetting = setting.Serialize().Deserialize<OverridableBindable<float>>();
+            Assert.That(deserializedSetting.HasCustomValue.Value, Is.True);
+            Assert.That(deserializedSetting.CustomValue.Value, Is.EqualTo(5f));
+
+            setting.HasCustomValue.Value = false;
+
+            Assert.That(setting.Serialize().Deserialize<float?>(), Is.Null);
```
Comment (bdach) [inline osu.Game.Tests/NonVisual/OverridableBindableTest.cs:150]: should have equivalent testing to the `HasCustomValue = true` case (deserialising to an `OverridableBindable<float>` should yield a non-null one with `HasCustomValue = true`)

File: osu.Game/Rulesets/Mods/OverridableBindable.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                    return;
+
+                case OverridableBindable<T> setting:
+                    if (setting.HasCustomValue.Value)
+                    {
+                        CustomValue.Value = setting.CustomValue.Value;
+                        HasCustomValue.Value = true;
+                    }
+                    else
+                        HasCustomValue.Value = false;
```
Comment (bdach) [inline osu.Game/Rulesets/Mods/OverridableBindable.cs:100]: can be simplified to
```suggestion
					HasCustomValue.Value = setting.CustomValue.Value;
                    if (setting.HasCustomValue.Value)
                        CustomValue.Value = setting.CustomValue.Value;
```

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Mods/OverridableBindable.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        public Bindable<T> FinalValue => finalValue;
+
+        /// <summary>
+        /// Whether the final bindable should be bound to <see cref="CustomValue"/>, otherwise will be immutably bound to <see cref="BaseValue"/>.
+        /// </summary>
+        public Bindable<bool> HasCustomValue { get; } = new BindableBool();
+
+        private readonly BindableNumber<T> immutableBaseValue = new BindableNumber<T>();
+        private readonly BindableNumber<T> customValue = new BindableNumber<T>();
+        private readonly BindableNumber<T> finalValue = new BindableNumber<T>();
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/OverridableBindable.cs:42]: good catch

File: osu.Game.Tests/NonVisual/OverridableBindableTest.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game.Tests/NonVisual/OverridableBindableTest.cs:150]: that is doing so right above this code

File: osu.Game/Rulesets/Mods/OverridableBindable.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/OverridableBindable.cs:100]: intentionally did it this way such that `FinalValue` binds to `CustomValue` (`HasCustomValue = true`) after the value is parsed.

Comment (bdach) [reply osu.Game/Rulesets/Mods/OverridableBindable.cs:100]: then reverse the operation order? with a comment, obviously. no idea why there was no comment in the first place if this was relying on weird operation ordering.

File: osu.Game.Tests/NonVisual/OverridableBindableTest.cs (hunk 2 above)
Comment (bdach) [reply osu.Game.Tests/NonVisual/OverridableBindableTest.cs:150]: i'm not seeing a

```csharp
            var deserializedSetting = setting.Serialize().Deserialize<OverridableBindable<float>>();
            Assert.That(deserializedSetting.HasCustomValue.Value, Is.False);
```

anywhere? i was explicitly saying that the deserialise/parse should also cover the no-custom-value case.

File: osu.Game/Rulesets/Mods/OverridableBindable.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/OverridableBindable.cs:100]: thought it was quite clear, not say "I have custom value" unless I actually do have a valid one, so you mean like this:
```suggestion
                    HasCustomValue.Value = false;

                    if (setting.HasCustomValue.Value)
                    {
                        CustomValue.Value = setting.CustomValue.Value;
                        HasCustomValue.Value = true;
					}
```

Comment (bdach) [reply osu.Game/Rulesets/Mods/OverridableBindable.cs:100]: why is the assignment to `HasCustomValue` under a branch at all if you're just copying the value, is what i'm getting at

File: osu.Game.Tests/NonVisual/OverridableBindableTest.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game.Tests/NonVisual/OverridableBindableTest.cs:150]: oh you mean it like that, sorry didn't fully understand your preivous comment

Comment (bdach) [reply osu.Game.Tests/NonVisual/OverridableBindableTest.cs:150]: i did flip true to false i suppose

File: osu.Game/Rulesets/Mods/OverridableBindable.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/OverridableBindable.cs]: Resolved in https://github.com/ppy/osu/pull/11387/commits/831e11dbf2a608bbbc5f061b846cc4a5caedcdf7 (should've been in a separate commit but /shrug)

**Comment by smoogipoo** [comment]
With the recent changes to the DA mod, I don't think this is relevant anymore.

**Pull request closed without merging**